docs/DafnyRef/Refinement.md

```diff
@@ -92,32 +92,133 @@ TODO
+definition to
+
+* provide a body missing in the parent,
+* replace the body of the parent with a new, semantically-compatible
+  body,
+* strengthen the postcondition of the parent method by adding one or more
+  `ensures` clauses,
+* provide a more deterministic version of a non-deterministic parent
+  body, or
+* amend the body of the parent method with additional statements.
+
+The type signature of a refining method must be the same as that of the
+parent method it refines. This can be ensured by providing an explicit
+type signature equivalent to that of the parent (with renaming of
+parameters allowed) or by using an ellipsis (`...`) to indicate copying
+of the parent type signature. The following example illustrates that
+latter case.
+
+```
+module M0 {
+  method m(x: int) returns (r: int)
+}
+
+module M1 refines M0 {
+  method m...
+  {
+    return x + 1;
+  }
+}
+```
+
+As an example of semantic compatibility and increased determinism,
+consider the following example.
+
+```
+module M0 {
+  method m(x: int) returns (r: int)
+    ensures r > x;
+  {
+    var y :| y > x;
+    return y;
+  }
+}
+
+module M1 refines M0 {
+  method m(x: int) returns (r: int)
+  {
+    return x + 1;
+  }
+}
+```
+
+Here, `m` ensures that its result will be greater than `x`, but the
+version in `M0` returns an arbitrary value satisfying that
+postcondition. The version of `m` in `M1` is deterministic, and is
+required to satisfy all of the postconditions that its parent method
+declares.
+
+To introduce additional statements, the child method can include
+ellipses within the body to stand in for portions of code from the
+parent body. Dafny then attempts to merge the body of the child with the
+body of the parent by filling in the ellipses. For example:
+
+```
+module M0 {
+  method m(x: int) returns (r: int)
+    ensures r > x;
+  {
+    var y :| y > x;
+    return y;
+  }
+}
+
+module M1 refines M0 {
+  method m(x: int) returns (r: int)
+  {
+    ...;
+    return y + 1;
+  }
+}
+```
+
+Refinement of method bodies can go far beyond adding statements to the
+end, however. Full details of the algorithm used to perform the merge
+operation are available in [this
+paper](https://dl.acm.org/doi/10.1007/s00165-012-0254-3).
+
+A refining module can also introduce new `method` declarations or
+definitions that do not exist in the refinement parent.
 
 ## 21.6. Lemma declarations
 
-TODO
+As lemmas are (ghost) methods, the description of method refinement from
+the previous section also applies to lemma refinement.
 
 ## 21.7. Function and predicate declarations
 
-TODO
+A child `function` (or `predicate`) definition can refine a parent
+`function` (or `predicate`) declaration or definition to
+
+* provide a body missing in the parent,
+* replace the body from the parent with a new, semantically-compatible
+  body, or
+* strengthen the postcondition of the parent function by adding one or more
+  `ensures` clauses.
```

Comment:
But also weaken the preconditions presumably?
I wonder if a general description of the relation between the parent and the child might not be simpler than exhaustive enumeration

Reply:
This is sound, but there is no notation for it. (This is done differently for `trait`s and for refinement modules. In a type that extends a trait, the entire specification as to be given again. This is verbose, but makes it clear what the specification is and gives the freedom to, for example, weaken preconditions. For refinement modules, there are so many things that are not shown explicitly in the child, and the basic idea is to just "superimpose" things onto the parent. Thus, you can *add* `ensures` clauses, but can't add `modifies` or `requires` clauses, and there is no way to *delete* a previously given specification clause.